### PR TITLE
#1 Pull Request for Messaging Unit Tests

### PR DESCRIPTION
**Pull request by charansingh83sh**
#1 Pull Request for Messaging Unit Tests

### CODE REVIEW

File: IO.Eventuate.Tram.Tests/Messages/MessageTest.cs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+            }
+        }
+        [TestMethod]
+        public void SendMessage()
+        {
+            IdGenerator generator = new IdGenerator();
+            var options = new DbContextOptionsBuilder<EventuateTramDbContext>().UseSqlServer(TestSettings.EventuateTramDbConnection).Options;
+            var schema = new EventuateSchema(TestSettings.EventuateTramDbSchema);
+            EventuateTramDbContextProvider provider = new EventuateTramDbContextProvider(options, schema);
+            MessageProducer mp = new MessageProducer(new List<IMessageInterceptor>(),
```
Comment (cer) [inline IO.Eventuate.Tram.Tests/Messages/MessageTest.cs:66]: `mp` should be the field `MessageTest.messageProducer` - not a local variable.

**Comment by cer** [comment]
@charansingh83sh where is the docker-compose.yml

**Comment by cer** [comment]
One other comment.
There should a single test method that subscribes(), sends a message, and verifies that the message was received.
It should look like this: 

https://github.com/eventuate-tram/eventuate-tram-core-examples-basic/blob/31832bb23ddc0c10458ca1044710d4c41d382c36/eventuate-tram-examples-common/src/main/java/io/eventuate/tram/examples/basic/messages/AbstractTramMessageTest.java#L34-L47

All of the initialization should be done in [TestInitialize] methods.

File: IO.Eventuate.Tram.Tests/Messages/MessageTest.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+    [TestClass]
+    public class MessageTest
+    {
+        public string uniqueId = System.DateTime.Now.Ticks.ToString();
+        public string subscriberId = "";
+        public string destination = "";
+        public string payload = "";
+        public DatabaseMessageProducer messageProducer;
+        private List<IO.Eventuate.Tram.Messaging.Common.Message> queue = new List<IO.Eventuate.Tram.Messaging.Common.Message>();
```
Comment (cer) [inline IO.Eventuate.Tram.Tests/Messages/MessageTest.cs:28]: The queue should be `List<IMessage>`

File: IO.Eventuate.Tram.Tests/Messages/MessageTest.cs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            generator, provider, Substitute.For<ILogger<DatabaseMessageProducer>>());
+            mp.Send(destination, MessageBuilder.WithPayload(payload).Build());
+
+            Assert.IsNotNull(destination);
+        }
+
+        [TestMethod]
+        public void SubscribeMessage()
+        {
+            PrePostHandlerMessageHandlerDecorator obj = new PrePostHandlerMessageHandlerDecorator();
```
Comment (cer) [inline IO.Eventuate.Tram.Tests/Messages/MessageTest.cs:76]: You can pass in an empty list of decorators.

File: IO.Eventuate.Tram.Tests/IO.Eventuate.Tram.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    <PackageReference Include="MSTest.TestFramework" Version="2.1.0" />
+    <PackageReference Include="coverlet.collector" Version="1.2.0" />
+    <PackageReference Include="NSubstitute" Version="4.2.2" />
+    <PackageReference Include="NUnit" Version="3.12.0" />
+    <PackageReference Include="NUnit3TestAdapter" Version="3.17.0" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Reference Include="IO.Eventuate.Tram">
+      <HintPath>bin\Debug\netcoreapp3.1\IO.Eventuate.Tram.dll</HintPath>
```
[the remark this answers is not in the document]
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/IO.Eventuate.Tram.Tests.csproj]: Reference Added

File: IO.Eventuate.Tram.Tests/Messages/MessageTest.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        public ISet<string> channels = new HashSet<string>();
+        public MessageTest()
+        {
+            uniqueId = System.DateTime.Now.Ticks.ToString();
+            subscriberId = "subscriberId" + uniqueId;
+            destination = "destination" + uniqueId;
+            payload = "Hello" + uniqueId;
+        }
+
+        public class MessageProducer : DatabaseMessageProducer
```
[the remark this answers is not in the document]
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/Messages/MessageTest.cs]: Class Removed.

File: IO.Eventuate.Tram.Tests/Messages/MessageTest.cs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        {
+            public MessageProducer(IEnumerable<IMessageInterceptor> messageInterceptors, IIdGenerator idGenerator, IEventuateTramDbContextProvider dbContextProvider, ILogger<DatabaseMessageProducer> logger)
+                : base(messageInterceptors, idGenerator, dbContextProvider,
+                logger)
+            {
+
+            }
+        }
+        public class MessageConsumer : KafkaMessageConsumer
```
[the remark this answers is not in the document]
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/Messages/MessageTest.cs]: Class Removed

File: IO.Eventuate.Tram.Tests/Messages/MessageTest.cs (hunk 1 above)
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/Messages/MessageTest.cs]: Done

**Comment by charansingh83sh** [comment]
> @charansingh83sh where is the docker-compose.yml

I have added docker-compose.yml with zookeeper and kafka services. Working on to resolve issue(_Login timeout expired. TCP Provider: Error code 0x2749 docker_) with mssql service. I will add mssql and cdc services once issue resolved.

File: IO.Eventuate.Tram.Tests/Messages/MessageTest.cs (hunk 2 above)
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/Messages/MessageTest.cs]: I have replaced the above list with _BlockingCollection<IMessage> queue = new BlockingCollection<IMessage>();_ to wait for message to receive.

File: IO.Eventuate.Tram.Tests/Messages/MessageTest.cs (hunk 3 above)
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/Messages/MessageTest.cs]: Getting below exception if I pass empty list of decorators.
_Test method IO.Eventuate.Tram.Tests.MessageTest.ShouldReceiveMessage threw exception: 
    System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')_

File: IO.Eventuate.Tram.Tests/Messages/MessageTest.cs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+{
+
+    [TestClass]
+    public class MessageTest
+    {
+        public string uniqueId = System.DateTime.Now.Ticks.ToString();
+        public string subscriberId = "";
+        public string destination = "";
+        public string payload = "";
+        public DatabaseMessageProducer messageProducer;
```
[the remark this answers is not in the document]
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/Messages/MessageTest.cs]: > One other comment.
> There should a single test method that subscribes(), sends a message, and verifies that the message was received.
> It should look like this:
> 
> https://github.com/eventuate-tram/eventuate-tram-core-examples-basic/blob/31832bb23ddc0c10458ca1044710d4c41d382c36/eventuate-tram-examples-common/src/main/java/io/eventuate/tram/examples/basic/messages/AbstractTramMessageTest.java#L34-L47
> 
> All of the initialization should be done in [TestInitialize] methods.

This is done.

File: IO.Eventuate.Tram.Tests/Messages/MessageTest.cs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+                loggerFactory, serviceScopeFactory)
+            {
+
+            }
+        }
+        [TestMethod]
+        public void SendMessage()
+        {
+            IdGenerator generator = new IdGenerator();
+            var options = new DbContextOptionsBuilder<EventuateTramDbContext>().UseSqlServer(TestSettings.EventuateTramDbConnection).Options;
```
[the remark this answers is not in the document]
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/Messages/MessageTest.cs:50]: Done

**Comment by cer** [comment]
> > @charansingh83sh where is the docker-compose.yml
> 
> I have added docker-compose.yml with zookeeper and kafka services. Working on to resolve issue(_Login timeout expired. TCP Provider: Error code 0x2749 docker_) with mssql service. I will add mssql and cdc services once issue resolved.

Please add those mssql and cdc service so I can see what the problem is?

Also, where is the error occurring? In the tests or in the cdc service? Please create issue describing problem.

File: IO.Eventuate.Tram.Tests/docker-compose.yml [hunk 9]
```diff
@@ -28,3 +38,31 @@ services:
...
+    image: eventuateio/eventuate-cdc-service:${CDC_SERVICE_DOCKER_VERSION:?You must set the CDC_SERVICE_DOCKER_VERSION environment variable.}
+    depends_on:
+    - kafka
+    - mssql
+    environment:
+      EVENTUATELOCAL_KAFKA_BOOTSTRAP_SERVERS: kafka:29092
+      EVENTUATELOCAL_ZOOKEEPER_CONNECTION_STRING: zookeeper:2181
+      EVENTUATE_CDC_READER_READER1_TYPE: polling
+      EVENTUATE_CDC_READER_READER1_DATASOURCEURL: jdbc:sqlserver:LAPTOP-7G6A2VOR;databaseName=TramDb
+      EVENTUATE_CDC_READER_READER1_DATASOURCEUSERNAME: sa
```
Comment (cer) [inline IO.Eventuate.Tram.Tests/docker-compose.yml:51]: This hostname is incorrect. `LAPTOP-7G6A2VOR` should be `mssql` - the name of the MS SQL server container

You just need to copy https://github.com/eventuate-tram/eventuate-tram-core-dotnet/blob/master/IO.Eventuate.Tram.IntegrationTests/docker-compose.yml

File: IO.Eventuate.Tram.Tests/docker-compose.yml [hunk 10]
```diff
@@ -1,6 +1,16 @@
 ---
 version: '3.4'
 services:
+  dbsetup:
+    image: mcr.microsoft.com/mssql-tools@sha256:62556500522072535cb3df2bb5965333dded9be47000473e9e0f84118e248642
+    depends_on:
+    - mssql
+    environment:
+      TRAM_DB_SERVER: "LAPTOP-7G6A2VOR"
```
Comment (cer) [inline IO.Eventuate.Tram.Tests/docker-compose.yml:9]: See other comment about hostname for mssql

File: IO.Eventuate.Tram.Tests/docker-compose.yml [hunk 11]
```diff
@@ -1,6 +1,16 @@
 ---
 version: '3.4'
 services:
+  dbsetup:
```
Comment (cer) [inline IO.Eventuate.Tram.Tests/docker-compose.yml:4]: PLease look at https://github.com/eventuate-tram/eventuate-tram-core-dotnet/blob/c7259f4e03345492253d5d8fe3af7d40edfe1262/IO.Eventuate.Tram.IntegrationTests/docker-compose.yml#L3

You just need to copy it.

In particular you need the `volumes` to point to the SQL files to initialize the database and you need add those SQL files to the repository

File: IO.Eventuate.Tram.Tests/docker-compose.yml (hunk 9 above)
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/docker-compose.yml:59]: Done

File: IO.Eventuate.Tram.Tests/docker-compose.yml (hunk 10 above)
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/docker-compose.yml]: Done

File: IO.Eventuate.Tram.Tests/docker-compose.yml (hunk 11 above)
Comment (charansingh83sh) [reply IO.Eventuate.Tram.Tests/docker-compose.yml:3]: Done

**Comment by cer** [comment]
Merged into branch wip-initial-version

**Pull request closed without merging**